### PR TITLE
FEA Implement an optimized sliding window matmul kernel

### PR DESCRIPTION
**Pull request by fcharras**
Necessary step before going forward with the KNN. This kernel will enable computing pairwise distances.

### CODE REVIEW

**Comment by ogrisel** [comment]
Nice, quick benchmark vs `np.dot` (backend by OpenBLAS SGEMM) on random `float32` arrays, e.g.:

```python
import numpy as np
n, m, k = 2048, 2048, 2048
dtype = np.float32

A = np.random.randn(n, k).astype(dtype)
B = np.random.randn(k, m).astype(dtype)
C = np.empty(shape=(n, m), dtype=dtype)
%timeit np.dot(A, B, out=C)
```

**Comment by fcharras** [comment]
Numpy
```
70 ms ± 2.2 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

sklearn_numba_dpex

```
158 ms ± 442 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

We're only twice slower :grin:

**Comment by ogrisel** [comment]
sycl-blas has some Python wrapper here: https://github.com/codeplaysoftware/sycl-blas/blob/master/python_generator/py_gen_blas_gemm_launcher.py

**Comment by ogrisel** [comment]
Actually this doesn't look like a wrapper. I am not sure what's the purpose of this code. Maybe a code generator.

**Comment by jjerphan** [comment]
Thanks for the benchmarks, @fcharras. :)

I would be in favor of reviewing and merging this PR if there's no downside: this implementation works, is efficient, and can be made even more efficient then if we want it to.

What do you think?

**Comment by fcharras** [comment]
Yes IMO it can be reviewed

**Comment by fcharras** [comment]
`dpnp.matmul` on GPU provides 10* speedup on numpy. And 20* speedup on the kernel in this PR. Will probably just use `dpnp.matmul` instead.

File: sklearn_numba_dpex/common/matmul.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+    compiled and fused into the first kernel as a device function with the help of
+    `dpex.func`. This comes with limitations as explained in:
+
+    https://intelpython.github.io/numba-dpex/latest/user_guides/kernel_programming_guide/device-functions.html # noqa
+
+    It is expected to take one scalar argument and return one scalar value.
+
+    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
+    multiply operator . It is expected to take two scalar argument and return one scalar
+    value.
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/matmul.py:43]:
```suggestion
    Likewise, if `multiply_fn` is not None it will be used in place of the scalar
    multiply operator. It is expected to take two scalar arguments and return one
    scalar value.
```

File: sklearn_numba_dpex/common/matmul.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+        group_first_col_idx = dpex.get_group_id(0) * square_window_side
+        result_col_idx = group_first_col_idx + local_col_idx
+
+        # Allocate shared memory for the two sliding windows on `X` and `Y_t`
+        X_sliding_window = dpex.local.array(shape=window_shape, dtype=dtype)
+        Y_t_sliding_window = dpex.local.array(shape=window_shape, dtype=dtype)
+
+        # Index of the column of the sliding window that the current work item will be
+        # responsible for loading. The "sliding" is materialize by incrementing this
+        # index by `square_window_side` at each iteration of the main loop.
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/matmul.py:144]:
```suggestion
        # Index of the column of the sliding window that the current work item will be
        # responsible for loading. The "sliding" is materialized by incrementing this
        # index by `square_window_side` at each iteration of the main loop.
```

File: sklearn_numba_dpex/common/matmul.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        window_col_idx = local_col_idx
+
+        # For both `X` and `Y_t`, that are assumed to be C-contiguous, in order to
+        # optimize RW memory access patterns, it's best to ensure that contiguous work
+        # items read contiguous memory slots in both shared and global memory. This is
+        # straightforward to ensure for windows on `X`, but requires the following
+        # remapped index for `Y_T`:
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/matmul.py:151]:
```suggestion
        # remapped index for `Y_t`:
```

File: sklearn_numba_dpex/common/matmul.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+                else:
+                    for idx in range(square_window_side):
+                        output += multiply_fn(
+                            X_sliding_window[local_row_idx, idx],
+                            Y_t_sliding_window[local_col_idx, idx],
+                        )
+
+            dpex.barrier(dpex.LOCAL_MEM_FENCE)
+
+        # Write result
```
Comment (jjerphan) [inline sklearn_numba_dpex/common/matmul.py:224]:
```suggestion
        # Save the value of the scalar product of the result_row_idx-th row of X
        # and of the result_col_idx-th column of Y.
```

**Review by jjerphan — COMMENTED** [review]
A few comments. Do we need to introduce this kernel since `dnpn.matmul` exist?

**Comment by fcharras** [comment]
> A few comments. Do we need to introduce this kernel since `dnpn.matmul` exist?

I wonder. I don't understand how the performance look so weak here, while our kmeans seems to work well, at least it compares as one could expect with cpu implementation.

I suppose the performances will not be as bad with typical input dimension for KNN (with `n_cols` = `n_features` is much lesser than the other two dimensions), and then it might actually be better to use this kernel because we can fuse `multiply_fn` and `elementwise_fn` to compute pairwise distances in one kernel call.

**Comment by fcharras** [comment]
I'm currently working on an update on this. I think it's actually possible to achieve near dpnp performance with numba_dpex.

**Comment by fcharras** [comment]
Also I've opened a discussion at `numba_dpex` see https://github.com/IntelPython/numba-dpex/issues/984

File: sklearn_numba_dpex/common/matmul.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+import numba_dpex as dpex
+import numpy as np
+
+from sklearn_numba_dpex.common._utils import _check_max_work_group_size
+
+zero_idx = np.int64(0)
+
+
+@lru_cache
+def make_matmul_2d_kernel(
```
[the remark this answers is not in the document]
Comment (ogrisel) [reply sklearn_numba_dpex/common/matmul.py:13]: I am not sure we need the full GEMM API in most cases and specializing it for simple MM might be more efficient for the most common cases (while keeping the code simpler).

I think a full GEMM kernel (or family of kernels to be specialized for subcases) can be done upstream in numba_dpex if needed.

File: sklearn_numba_dpex/common/matmul.py [hunk 6]
```diff
@@ -0,0 +1,814 @@
...
+        work_item_row_idx = work_item_idx // sub_group_size
+        work_item_col_idx = work_item_idx % sub_group_size
+
+        group_idx = dpex.get_group_id(zero_idx)
+
+        # Get indices of the row and the column of the top-left corner of the sub-array
+        # of results covered by this work group
+        group_row_idx = group_idx // global_grid_n_cols
+        group_col_idx = group_idx % global_grid_n_cols
+        # TODO: improve cache locality ?
```
Comment (ogrisel) [inline sklearn_numba_dpex/common/matmul.py:344]: nit

```suggestion
        # TODO: improve cache locality?
```

File: sklearn_numba_dpex/common/matmul.py [hunk 7]
```diff
@@ -0,0 +1,814 @@
...
+    # `arithmetic_intensity_multiplier_X + arithmetic_intensity_multiplier_Y`
+    # more items in the work group shared memory, per item in this work group.
+
+    # Those values should be set high enough so that the kernel is compute-bound
+    # rather than memory-bound.
+
+    # Setting it too high might pressure local or private memory too much for it to be
+    # worth.
+    arithmetic_intensity_multiplier_X = 4
+    arithmetic_intensity_multiplier_Y = 2
```
Comment (ogrisel) [inline sklearn_numba_dpex/common/matmul.py:112]: Let's make those kwargs of `make_matmul_2d_kernel` with the same default values so that it will be possible in the future to do run some kind of auto-tuning grid search.

**Review by ogrisel — APPROVED** [review]
I won't review the core of the kernel itself but +1 for merge with stronger non-regression tests as explained below (at least a few more test cases with larger shapes and maybe non-integer, positive and negative input values).

File: sklearn_numba_dpex/common/tests/test_matmul.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        ((5, 8), (8, 4)),
+        ((5, 10), (10, 4)),
+    ],
+)
+@pytest.mark.parametrize("dtype", float_dtype_params)
+def test_matmul_2d(test_input_shapes, work_group_size, sub_group_size, dtype):
+    X_shape, Y_shape = test_input_shapes
+
+    X = _arange_reshaped(X_shape, dtype)
+    Y = _arange_reshaped(Y_shape, dtype)
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/common/tests/test_matmul.py:52]: It doesn't cost much more to have both the `_arange_reshaped` and `random` tests so I can add both. (I think `_arange_reshaped` is enough, but above that it is handier for debug than random values so I'd like to keep it along)

I think the boundary effects are already all covered by this set of test cases (the inputs are smalls but the tests also use small values for `work_group_size` / `sub_group_size`)

**Comment by fcharras** [comment]
Latest push should have addressed comments, this version is mergeable IMO.

- [x] improved tests
- [x] removed WIP work on cache locality (and pushed it in the matmul_kernel_wip on this repo)
- [x] exposed all performance parameters

**Review by ogrisel — APPROVED** [review]
